src/bvals/cc/bvals_cc_in_one.cpp

```diff
@@ -0,0 +1,553 @@
+//         and to coarser.
+
+void CalcIndicesLoadSame(int ox, int &s, int &e, const IndexRange &bounds) {
+  if (ox == 0) {
+    s = bounds.s;
+    e = bounds.e;
+  } else if (ox > 0) {
+    s = bounds.e - NGHOST + 1;
+    e = bounds.e;
+  } else {
+    s = bounds.s;
+    e = bounds.s + NGHOST - 1;
+  }
+}
+
+//----------------------------------------------------------------------------------------
+//! \fn void cell_centered_bvars::CalcIndicesLoadToFiner(int &si, int &ei, int &sj,
+//                                                       int &ej, int &sk, int &ek,
+//                                                       const NeighborBlock &nb,
+//                                                       MeshBlock *pmb)
+//  \brief Calculate indices for LoadBoundary routines for buffers to finer levels
+
+void CalcIndicesLoadToFiner(int &si, int &ei, int &sj, int &ej, int &sk, int &ek,
+                            const NeighborBlock &nb, MeshBlock *pmb) {
+  int cn = pmb->cnghost - 1;
+
+  IndexDomain interior = IndexDomain::interior;
+  const IndexShape &cellbounds = pmb->cellbounds;
+  si = (nb.ni.ox1 > 0) ? (cellbounds.ie(interior) - cn) : cellbounds.is(interior);
+  ei = (nb.ni.ox1 < 0) ? (cellbounds.is(interior) + cn) : cellbounds.ie(interior);
+  sj = (nb.ni.ox2 > 0) ? (cellbounds.je(interior) - cn) : cellbounds.js(interior);
+  ej = (nb.ni.ox2 < 0) ? (cellbounds.js(interior) + cn) : cellbounds.je(interior);
+  sk = (nb.ni.ox3 > 0) ? (cellbounds.ke(interior) - cn) : cellbounds.ks(interior);
+  ek = (nb.ni.ox3 < 0) ? (cellbounds.ks(interior) + cn) : cellbounds.ke(interior);
+
+  // send the data first and later prolongate on the target block
+  // need to add edges for faces, add corners for edges
+  if (nb.ni.ox1 == 0) {
+    if (nb.ni.fi1 == 1)
+      si += pmb->block_size.nx1 / 2 - pmb->cnghost;
+    else
+      ei -= pmb->block_size.nx1 / 2 - pmb->cnghost;
+  }
+  if (nb.ni.ox2 == 0 && pmb->block_size.nx2 > 1) {
+    if (nb.ni.ox1 != 0) {
+      if (nb.ni.fi1 == 1)
+        sj += pmb->block_size.nx2 / 2 - pmb->cnghost;
+      else
+        ej -= pmb->block_size.nx2 / 2 - pmb->cnghost;
+    } else {
+      if (nb.ni.fi2 == 1)
+        sj += pmb->block_size.nx2 / 2 - pmb->cnghost;
+      else
+        ej -= pmb->block_size.nx2 / 2 - pmb->cnghost;
+    }
+  }
+  if (nb.ni.ox3 == 0 && pmb->block_size.nx3 > 1) {
+    if (nb.ni.ox1 != 0 && nb.ni.ox2 != 0) {
+      if (nb.ni.fi1 == 1)
+        sk += pmb->block_size.nx3 / 2 - pmb->cnghost;
+      else
+        ek -= pmb->block_size.nx3 / 2 - pmb->cnghost;
+    } else {
+      if (nb.ni.fi2 == 1)
+        sk += pmb->block_size.nx3 / 2 - pmb->cnghost;
+      else
+        ek -= pmb->block_size.nx3 / 2 - pmb->cnghost;
+    }
+  }
+}
+
+// send boundary buffers with MeshBlockPack support
+// TODO(pgrete) should probaly be moved to the bvals or interface folders
+TaskStatus SendBoundaryBuffers(std::shared_ptr<MeshData<Real>> &md) {
```

Comment:
This function is long and kind of hard to follow. It might be nice to break some pieces into sub-functions. For example the kokkos kernels have a lot of boiler plate.

Reply:
Yes, this could be split into three dependent tasks.
I kept it in one (similar to the load balance function) as the subtasks are highly dependent (i.e., there'd be a linear dependency between those where no other task should interrupt), but I'm also happy to split it.

Reply:
we could also keep it a single task, but split out "subroutines" in the file just to give an easier picture of flow of the function.